content/en/blog/_posts/non-root-containers-and-devices.md

```diff
@@ -0,0 +1,222 @@
+---
+layout: blog
+title: 'Non-root Containers And Devices'
+date: 2021-09-YY
+slug: non-root-containers-and-devices
+---
+
+**Author:** Mikko Ylinen (Intel)
+
+The user/group ID related security settings in Pod's `securityContext` trigger a problem when users want to
+deploy containers that use accelerator devices (via [Kubernetes Device Plugins](/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/)) on Linux. In this blog
+post I talk about the problem and describe the work done so far to address it. It's not meant to be a long story about getting the [k/k issue](https://github.com/kubernetes/kubernetes/issues/92211) fixed.
+
+Instead, this post aims to raise awareness of the issue and to highlight important device use-cases too. This is needed as Kubernetes works on new related features such as support for user namespaces.
+
+## Why non-root containers can't use devices and why it matters
+One of the key security principles for running containers in Kubernetes is the
+principle of least privilege. The Pod/container `securityContext` specifies the config
+options to set, e.g., Linux capabilities, MAC policies, and user/group ID values to achieve this.
+
+Furthermore, the cluster admins are supported with tools like [PodSecurityPolicy](/docs/concepts/policy/pod-security-policy/) (deprecated) or
+[Pod Security Admission](/docs/concepts/security/pod-security-admission/) (alpha) to enforce the desired security settings for pods that are being deployed in
+the cluster. These settings could, for instance, require that containers must be `runAsNonRoot` or
+that they are forbidden from running with root's group ID in `runAsGroup` or `supplementalGroups`.
+
+In Kubernetes, the kubelet builds the list of [`Device`](https://pkg.go.dev/k8s.io/cri-api@v0.22.1/pkg/apis/runtime/v1#Device) resources to be made available to a container
+(based on inputs from the Device Plugins) and the list is included in the CreateContainer CRI message
+sent to the CRI container runtime. Each `Device` contains little information: host/container device
+paths and the desired devices cgroups permissions.
+
+The [OCI Runtime Spec for Linux Container Configuration](https://github.com/opencontainers/runtime-spec/blob/master/config-linux.md)
+expects that in addition to the devices cgroup fields, more detailed information about the devices
+must be provided:
+
+```yaml
+{
+        "type": "<string>",
+        "path": "<string>",
+        "major": <int64>,
+        "minor": <int64>,
+        "fileMode": <uint32>,
+        "uid": <uint32>,
+        "gid": <uint32>
+},
+```
+
+The CRI container runtimes (containerd, CRI-O) are responsible for obtaining this information
+from the host for each `Device`. By default, the runtimes copy the host device's user and group IDs:
+
+- `uid` (uint32, OPTIONAL) - id of device owner in the container namespace.  
+- `gid` (uint32, OPTIONAL) - id of device group in the container namespace.
```

Comment:
I might omit namespace here. Does this technology require that the container runtime is one that uses (Linux) namespaces? If not, let's be more runtime-agnostic.

Reply:
This is quoting the official OCI spec: https://github.com/opencontainers/runtime-spec/blob/master/config-linux.md#devices